Select: Convert to functional component

westwood846 opened the pull request:
Fixes #251.

rosslo commented on src/Input/Select/Select.tsx:
```diff
@@ -15,346 +15,271 @@ import {
+  const handleSelect = React.useCallback(() => {
+    const activeElement = getActiveElement();
+    const selectedValue = activeElement.innerText;
+    setSelectedValue(selectedValue || '');
```
Should we have an empty string as a fallback?

westwood846 replied:
Yeah this seems to have no effect, I'm removing it.

rosslo commented on src/Input/Select/Select.tsx:
```diff
@@ -15,346 +15,271 @@ import {
+        scrollToActiveElement();
+      } else if (e.keyCode === 40 && cursor < filterValue.length - 1) {
+        setCursor(cursor + 1);
+        scrollToActiveElement();
```
Saw the original usage is to call `scrollToActiveElement` after `cursor` was updated.

But the current one seems to call them concurrently.

westwood846 replied:
Good catch, I wasn't familiar with the second `setState` argument and didn't realise that this should be sequential. I added an `effect` that triggers the scroll _after_ `cursor` updates.

rosslo commented:
@westwood846 Could you rebase this PR onto the latest staging? Since I saw this PR uses some hooks which might cause the errors of `react-hooks/rules-of-hooks`. 🙏

rosslo commented on src/Input/Select/Select.tsx:
```diff
@@ -15,346 +15,271 @@ import {
+    const activeChild = filterValue[activeElementIndex];
+    const onOptionClick = get(activeChild, 'props.onOptionClick');
+    if (onOptionClick !== undefined) {
+      onOptionClick(selectedValue);
```
I saw we were passing `event` as the parameter of `onOptionClick`. 

Did you check the usage of it in the `marketplace` and `employers` projects?

westwood846 replied:
Re-added it, shouldn't have been removed.

rosslo commented on src/Input/Select/Select.tsx:
```diff
@@ -15,346 +15,271 @@ import {
-      });
-
-      if (onChange !== undefined) {
-        return onChange(e);
```
same as https://github.com/glints-dev/glints-aries/pull/252#discussion_r393438409

westwood846 replied:
The funny thing is, implementations seem to depend on the changed value being returned here. For example here: 
![image](https://user-images.githubusercontent.com/3341028/77391478-ce023680-6dd3-11ea-8e3c-dca0a9c78170.png) (app/modules/UgcQnA/Components/QuestionForm/TopicField.js in DST)

rosslo replied:
Yes, it will be handy if onChange and onOptionClick can return the changed value only.

Do you think we should refactor the use of Select on dst and employers for ^?

westwood846 replied:
I tested agains some sample DST usages. I think the current state of the PR preserves the previous behaviour. As evidenced by the screenshot above, the `onChange` component did return the changed value.

westwood846 commented:
> @westwood846 Could you rebase this PR onto the latest staging? Since I saw this PR uses some hooks which might cause the errors of `react-hooks/rules-of-hooks`. 🙏

Rebased and fixed!

rosslo commented:
Bug: Use the keyboard to select option, the floating label doesn't work well.
![image](https://user-images.githubusercontent.com/13249748/77400462-2d694200-6de6-11ea-980b-3d63987add72.png)

rosslo commented on src/Input/Select/Select.tsx:
```diff
@@ -15,346 +15,287 @@ import {
+  );
+
+  React.useEffect(() => {
+    if (cursor && shouldScrollToCursor) {
```
I think there must be a condition is `cursor` is zero and `shouldScrollToCursor ` is true. You can test with long options, you won't be able to scroll to first options.

westwood846 replied:
Removed!

rosslo commented on src/Input/Select/Select.tsx:
```diff
@@ -15,346 +15,287 @@ import {
+    (e?: void) => {
+      const activeElement = getActiveElement();
+      const selectedValue = activeElement.innerText;
+      // Need  || '' here to ensure that the select input is never uncontrolled,
```
Hmm, when will `selectedValue` be nil?

If it is defensive programming, I think we can just make `<SelectInput value={selectedValue || ''} // ... />` instead, then it might be safer.

westwood846 replied:
No this is some weird issue with React. Because you cannot change from uncontrolled to controlled component. `selectedValue` is undefined at some point before the user types, so the underlying select input would be interpreted as uncontrolled first. This simply makes sure that that never happens.

rosslo replied:
Yeah, I knew the point for controlled component, but after I played around with `Select` I still cannot understand how `selectedValue` can be `undefined` 🤔

westwood846 replied:
It happens in the tests. There are seven test cases were I could find this issue. They are the ones who use either `setupSelectOptionFromMenuWithKeyHandling` or `setupSelectOptionFromMenu`. Not quite sure why this is.

westwood846 replied:
Okay I found out that the `innerText` is sometimes undefined. `textContent` works better here. That's what's used by the tests too.

westwood846 replied:
> Bug: Use the keyboard to select option, the floating label doesn't work well.
> Btw, would you mind adding a test case to catch this haha?
> ![image](https://user-images.githubusercontent.com/13249748/77400462-2d694200-6de6-11ea-980b-3d63987add72.png)

Fixed

westwood846 commented:
>Bug: Use the keyboard to select option, the floating label doesn't work well.
>Btw, would you mind adding a test case to catch this haha?

Fixed the bug and added the test. I couldn't figure out how to test a component with snapshots after interaction so I'm just manually checking the `transform` style here. Not very elegant but gets the job done since the test will fail if any inadvertent changes are made to the floating behaviour.

rosslo commented on src/Input/Select/Select.test.tsx:
```diff
@@ -211,6 +219,12 @@ describe('when the enter key is pressed on an option', () => {
+  it('should float the label ', () => {
+    const { selectLabel } = setupSelectOptionFromMenuWithKeyHandling();
+    const style = window.getComputedStyle(selectLabel);
+    expect(style.transform).not.toEqual('');
```
I think we can use `toHaveStyle` instead.
```
expect(selectLabel).toHaveStyle('transform: translate3d(-15px,-20px,0);');
```

westwood846 replied:
Thanks for the tip! Implemented this :)

rosslo commented on src/Input/Select/Select.tsx:
```diff
@@ -15,346 +15,283 @@ import {
+  const handleSelect = React.useCallback(
+    (e?: React.ChangeEvent<HTMLInputElement>) => {
+      const activeElement = getActiveElement();
+      // const selectedValue = get(e, 'target.value', activeElement.textContent);
```
It seems you forgot to remove it.

westwood846 replied:
Yep! ^^
